build.py

```diff
@@ -0,0 +1,228 @@
+    '''
+    Use CMake to generate build files
+    '''
+    system = platform.system().lower()
+    cmake_generate_args = [
+        'cmake',
+        '-DCMAKE_INSTALL_PREFIX=' + os.path.join(build_dir(args), 'install')]
+    cmake_generate_env = os.environ.copy()
+    if 'windows' == system:
+        if 'x64' == args.architecture:
+            cmake_generate_args.extend(['-A', 'x64'])
+    else:
+        if 'debug' == args.configuration:
+            cmake_generate_args.append('-DCMAKE_BUILD_TYPE=Debug')
+        else:
+            cmake_generate_args.append('-DCMAKE_BUILD_TYPE=Release')
+        if (shutil.which('clang') is not None) and\
+                (shutil.which('clang++') is not None):
+            cmake_generate_env['CC'] = 'clang'
+            cmake_generate_env['CXX'] = 'clang++'
+    for config, dir in BUILD_CONFIGS.items():
+        for output in [('ARCHIVE', 'lib'), ('LIBRARY', 'bin'), ('RUNTIME', 'bin')]:
+            cmake_generate_args.append(
+                '-DCMAKE_{0}_OUTPUT_DIRECTORY_{1}={2}/{3}/{4}/{5}/{6}'.format(
+                    output[0],
+                    config.upper(),
+                    BUILD_ROOT,
+                    dir,
+                    system,
+                    args.architecture,
+                    output[1]))
+    cmake_generate_args.extend(cmake_generate_options(args))
+    work_dir = BUILD_ROOT
+    if(cmake_version() < CMAKE_VERSION_3_13):
+        work_dir = build_dir(args)
+        cmake_generate_args.append(BUILD_ROOT)
+    else:
+        cmake_generate_args.extend(['-S', '.', '-B', build_dir(args)])
+    os.makedirs(work_dir, mode=0o744, exist_ok=True)
+    cmake_generate_result = subprocess.run(
+        cmake_generate_args, cwd=work_dir, env=cmake_generate_env)
+    if 0 != cmake_generate_result.returncode:
+        raise BuildError('failed to generate build files')
+
+
+def cmake_build(args):
+    '''
+    Build using CMake
+    '''
+    cmake_build_args = ['cmake', '--build', '.']
+    if is_windows():
+        cmake_build_args.extend(
+            ['--config', args.configuration.capitalize()])
+    if args.clean or args.clobber:
+        cmake_build_args.extend(['--target', 'clean'])
+    cmake_build_result = subprocess.run(
+        cmake_build_args, cwd=build_dir(args))
+    if 0 != cmake_build_result.returncode:
+        raise BuildError('cmake build failed')
+
+
+# Main entry point
+if '__main__' == __name__:
+    try:
+        args = parse_args()
+        clean = args.clean or args.clobber
+        if not clean:
+            update_external_dependencies(args)
+        build_dir_exists = os.path.exists(build_dir(args))
+        if (clean and build_dir_exists) or (not clean):
+            cmake_generate_build_files(args)
+            cmake_build(args)
+        if args.clobber and build_dir_exists:
+            shutil.rmtree(build_dir(args))
+    except Exception as error:
+        print('Error', *(error.args))
+        sys.exit(1)
+    sys.exit(0)
```

Comment:
There are a few `pylint3` warnings in this file.  They're minor and don't impact functionality, but as a standard code style, it would be good to fix these.  I saw:
```
        ************* Module build
        C: 67, 0: Wrong hanging indentation (remove 4 spaces).
                        ', '.join(ARCHITECTURES)))
                    |   ^ (bad-continuation)
        C:184, 0: Unnecessary parens after 'if' keyword (superfluous-parens)
        C: 34,11: Comparison should be platform.system().lower() == 'windows' (misplaced-comparison-constant)
        W: 95,33: Redefining name 'args' from outer scope (line 215) (redefined-outer-name)
        C:102,11: Comparison should be update_git_submodule_result.returncode != 0 (misplaced-comparison-constant)
        W:106,14: Redefining name 'args' from outer scope (line 215) (redefined-outer-name)
        W:126, 4: Redefining name 'cmake_version' from outer scope (line 116) (redefined-outer-name)
        W:130,27: Redefining name 'args' from outer scope (line 215) (redefined-outer-name)
        W:150,31: Redefining name 'args' from outer scope (line 215) (redefined-outer-name)
        W:171,16: Redefining built-in 'dir' (redefined-builtin)
        C:159, 7: Comparison should be system == 'windows' (misplaced-comparison-constant)
        C:160,11: Comparison should be args.architecture == 'x64' (misplaced-comparison-constant)
        C:163,11: Comparison should be args.configuration == 'debug' (misplaced-comparison-constant)
        C:192, 7: Comparison should be cmake_generate_result.returncode != 0 (misplaced-comparison-constant)
        W:196,16: Redefining name 'args' from outer scope (line 215) (redefined-outer-name)
        C:208, 7: Comparison should be cmake_build_result.returncode != 0 (misplaced-comparison-constant)
        C:213, 3: Comparison should be __name__ == '__main__' (misplaced-comparison-constant)
        W:225,11: Catching too general exception Exception (broad-except)
        C:215, 8: Constant name "args" doesn't conform to UPPER_CASE naming style (invalid-name)
        C:216, 8: Constant name "clean" doesn't conform to UPPER_CASE naming style (invalid-name)
        C:219, 8: Constant name "build_dir_exists" doesn't conform to UPPER_CASE naming style (invalid-name)

        -----------------------------------
        Your code has been rated at 8.04/10
```